src/micro_error.c

```diff
@@ -0,0 +1,235 @@
+    va_end(args2);
+    return err;
+}
+
+microError *
+micro_Errorf(const char *format, ...)
+{
+    microError *err = NULL;
+    va_list args;
+
+    va_start(args, format);
+    err = verrorf(NATS_OK, 0, format, args);
+    va_end(args);
+    return err;
+}
+
+microError *
+micro_ErrorfCode(int code, const char *format, ...)
+{
+    microError *err = NULL;
+    va_list args;
+
+    va_start(args, format);
+    err = verrorf(NATS_OK, code, format, args);
+    va_end(args);
+    return err;
+}
+
+microError *
+micro_ErrorFromStatus(natsStatus s)
+{
+    microError *err = NULL;
+    const char *message = natsStatus_GetText(s);
+    size_t message_len = strlen(message);
+
+    if (s == NATS_OK)
+        return NULL;
+
+    err = NATS_CALLOC(1, sizeof(microError) + message_len + 1);
+    if (err == NULL)
+        return &_errorOutOfMemory;
+
+    err->status = s;
+    memcpy(err->message, message, message_len + 1);
+    return err;
+}
+
+microError *
+micro_is_error_message(natsStatus status, natsMsg *msg)
+{
+    microError *err = NULL;
+    const char *c = NULL, *d = NULL;
+    bool is_service_error;
+    bool is_nats_error = (status != NATS_OK);
+    int code = 0;
+
+    if (msg != NULL)
+    {
+        natsMsgHeader_Get(msg, MICRO_ERROR_CODE_HDR, &c);
+        natsMsgHeader_Get(msg, MICRO_ERROR_HDR, &d);
+    }
+    if (!nats_IsStringEmpty(c))
+    {
+        code = atoi(c);
+    }
+    is_service_error = (code != 0) || !nats_IsStringEmpty(d);
+    
+    if (is_service_error && !is_nats_error)
+    {
+        return micro_ErrorfCode(code, d);
+    }
+    else if (!is_service_error && is_nats_error)
+    {
+        return micro_ErrorFromStatus(status);
+    }
+    else if (is_service_error && is_nats_error)
+    {
+        err = microError_Wrapf(micro_ErrorFromStatus(status), d);
+        err->code = code;
+        return err;
+    }
+
+    return NULL;
+}
+
+microError *
+microError_Wrapf(microError *err, const char *format, ...)
+{
+    va_list args;
+    microError *new_err = NULL;
+
+    if (err == NULL)
+        return NULL;
+
+    va_start(args, format);
+    new_err = verrorf(NATS_OK, 0, format, args);
+    va_end(args);
+
+    new_err->cause = err;
+    return new_err;
+}
+
+const char *
+microError_String(microError *err, char *buf, int size)
+{
+    size_t used = 0;
+    const char *caused;
+
+    if (buf == NULL)
+        return "";
+    if (err == NULL)
+    {
+        snprintf(buf, size, "null");
+        return buf;
+    }
+
+    if (err->status != NATS_OK)
+    {
+        used += snprintf(buf + used, size - used, "status %d: ", err->status);
+    }
+    if (err->code != 0)
+    {
+        used += snprintf(buf + used, size - used, "code %d: ", err->code);
+    }
+    used += snprintf(buf + used, size - used, "%s", err->message);
+
+    if (err->cause != NULL)
+    {
+        used += snprintf(buf + used, size - used, ": ");
+        caused = microError_String(err->cause, buf + used, size - used);
```

Comment:
But then you have the warning here now since `microError_String` expects a `int` and your are passing a `size_t` (size-used):
```
src\micro_error.c(202): warning C4267: 'function': conversion from 'size_t' to 'int', possible loss of data
```